### PR TITLE
[WIP] Adding Python 3.x support

### PR DESCRIPTION
**Pull request by bcipolli**
This PR depends on two unmerged PRs (#369, #371), but I am opening this PR now for three reasons:

1. Supporting Python 2 and 3 takes creating and following compatibility conventions.  I wanted to list what I've discovered from the code so far, and open discussion to agreeing on such conventions.
2. To get any help (none needed, but appreciated) to solve remaining issues
3. To open early review to anybody who wishes to do so (see https://github.com/bcipolli/nilearn/compare/bcipolli:fix_warnings...python-3.x to view a set of commits specific to Python 2/3 support).
.

Open issues to solve (see https://travis-ci.org/bcipolli/nilearn/builds for latest branch builds / test errors):
* File-based string operations with numpy are really challenging in Python 3.x, because of changes in defaults (all strings are unicode).

Here's the current list of conventions I'm using so far:

**Python 2 conventions that are now Python 3 requirements**
* All `nilearn` imports should be relative (except, perhaps, in tests)
* Write `except Exception, e` as `except Exception as e`
* Write `isinstance(x, types.ListType)` as `isinstance(x, list)`
* Write `callable(x)` as `isinstance(x, collections.Callable)`
* Write `operator.isSequenceType(x)` as `isinstance(x, collections.Sequence)`
* Write `operator.isNumberType(x)` as `isinstance(x, collections.Number)`
* Write `nose.tools.assert_equals` as `nose.tools.assert_equal`
* Write `nilearn._utils.testing.assert_raises_regexp` as `nilearn._utils.testing.assert_raises_regex` (mirrors nose clarification)
* Write `<Iterable>.next()` as `next(<Iterable>)`
* Write `filename = os.tempnam()` as `(_, filename) = tempfile.mkstemp()`
* Write `not key in dict` as `key not in dict` (this actually warns in Python 3.x!)
* Make sure to use `with open('file') as f' syntax; any file pointers that are not used in a context manager nor explicitly closed throw warnings.

**Python 3 changes that require care to be backwards compatible**
* Non-unicode strings must be prefixed with `b`: `b'string'`(mostly an issue interacting with numpy, but also some low-level operations like gzip code)
* Write `print "%s" % (message,)` as `print("%s" % (message,))`
* Write integer division `3 / 2 == 1` as `3 // 2 == 1`
* Write `<dict>.iteritems()` as `<dict>.items()` (there is no iteritems in Python 3)
* Write `<dict>.itervalues()` as `<dict>.values()` (there is no itervalues in Python 3)
* Write `itertools.izip()` as `zip()` (there is no `izip` in Python 3)
* Write octal string `0100` `0o100`
* Write `round` as `np.round` (avoids a numpy issue in Python 3)

**Importing things from `six` instead of places you're used to**
* Write `basestring` as `six.string_types` (requires `import six`)
* Use the new urllib object structure instead of ulrllib/urllib2
    * import from `six.moves.urllib`.
    * there are no base-level functions/objects, all are present in subpackages
* Import  `StringIO` and `cPickle from `six.moves`

**Changes in how you design / write code**
* Think carefully about any use of `xrange`.  If you *require* an iterator, then you must import `past.builtins.xrange`.  Otherwise, just use `range` (which is a list in Python 2, and an interator in Python 3).
* Most things that used to generate lists now generate iterators; any time you need a list, make sure to wrap it in `list()`
    * **THIS WILL GET YOU!**
    * Includes:
        * `dict.keys()`, `dict.values()`
        * `zip`
        * `range`
        * `map`

### CODE REVIEW

**Comment by bcipolli** [comment]
I read up on most of the rest of the test failures--they are due to string handling differences between Python 2 and 3 and heavy usage in `nilearn` of numpy's `loadtxt` and `recfromtxt` functions without specifying column dtypes, and `nilearn`.

`numpy` has chosen to always return `loadtxt` and `recfromtxt` strings without types as bytes.  However, string literal defaults in Python are version dependent: Python 2 defaults strings to bytes, and Python 3 defaults to unicode.

This fails when in `nilearn` we explicitly filter based on condition masks using string literals, e.g. 

```python
haxby_dataset = datasets.fetch_haxby_simple()
conditions = np.recfromtxt(haxby_dataset.conditions_target)['f0']
condition_mask = np.logical_or(conditions == 'face', conditions == 'house')
```

This will succeed in Python 2, but fail in Python 3.  In Python 2, the `f0` column is bytes (default), the condition mask `'face'` is bytes, so the comparison succeeds.  In Python 3, the `f0` column is bytes (default), the condition mask `'face'` is unicode, so the comparison fails.

One solution could be to specify all string literals as bytes, e.g. `b'face'`.  This would work, but isn't good for examples (so few people would be aware of that.

Another solution could be to write our own `loadtxt` and `recfromtxt` functions that call `numpy`'s version, then convert any string fields to unicode by calling `decode()` on the strings.  I dislike wrapper functions like this because those familiar with `numpy` may no longer read the examples as clearly (similar, but different function is called....)

A final solution is to repackage the dataset code (where most of this info lies) more carefully, such that the column definitions and datatypes are part of the package.  I think this is the best option--we have that information (we spent the time to curate these datasets), and it would allow better column names (we could use `condition` instead of defaults like `f0`), but it would take some thinking about how to package that information well.

My suggestion: implement our own wrappers for `loadtxt` and `recfromtxt` and use them in our examples.  If there's interest in the third one, I can open an issue for it.

If no objections, I'll start on solution #2 on Tuesday.  Thanks!

**Comment by GaelVaroquaux** [comment]
If it's only on the examples, let's punt: the examples won't work out of
the box in Python 3. That's too bad. But given that a very large fraction
of our target users are not in Python 3, it will not affect them. On the
other hand, making the examples harder to read would affect them.

**Comment by bcipolli** [comment]
The issue also exists in `test_datasets.py`, but I don't think it's needed for any core code.  I assume with the tests, I can solve that either way (literals or through a conversion function).

**Comment by GaelVaroquaux** [comment]
> I assume with the tests, I can solve that either way (literals or
> through a conversion function).

I'd solve it via the simplest way, which probably does not entail writing
a wrapper function.

**Comment by bcipolli** [comment]
If punting, I suggest adding information somewhere (could be at the top of relevant examples, could be somewhere on the website) about this issue.  The examples are a core part of getting people on `nilearn`, and if claiming `python 3` support I think it's important to let users know that the examples won't work for them.

**Comment by GaelVaroquaux** [comment]
> If punting, I suggest adding information somewhere (could be at the top of
> relevant examples, could be somewhere on the website) about this issue.

Sounds right. If there aren't too many examples where this is the case, I
would add it there.

**Comment by lesteve** [comment]
It looks like using conditions == b'face' will work for both Python 2 and 3, i.e.:

```python
condition_mask = np.logical_or(conditions == 'face', conditions == 'house')
```

and another thing about:
> All nilearn imports should be relative (except, perhaps, in tests)

I seem to remember the scikit-learn policy was to use absolute import in tests. I guess this way you can easily run the test file independently from the package it is coming from, say run a test from your scikit-learn git checkout with an older version of scikit-learn for example.

File: README.rst [hunk 1]
```diff
@@ -32,6 +32,7 @@ The required dependencies to use the software are:
 * SciPy >= 0.7
 * Scikit-learn >= 0.12.1
 * Nibabel >= 1.1.0.
+* Six
```
Comment (lesteve) [inline README.rst]: I was thinking we could use the 'six' version that is shipped in sklearn.externals.

File: doc/sphinxext/gen_rst.py [hunk 2]
```diff
@@ -8,19 +8,19 @@
 Files that generate images should start with 'plot'
```
Comment (lesteve) [inline doc/sphinxext/gen_rst.py]: I don't think you should bother about this file too much. We are supposed to switch to [sphinx-gallery](http://github.com/sphinx-gallery/sphinx-gallery) at one point and I think it is already compatible with Python3.
